### PR TITLE
KAFKA-3824: Clarify autocommit delivery semantics for consumer

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 1]
```diff
@@ -65,55 +65,54 @@
 import java.util.regex.Pattern;
 
 /**
- * A Kafka client that consumes records from a Kafka cluster.
+ * A client that consumes records from a Kafka cluster.
  * <p>
- * It will transparently handle the failure of servers in the Kafka cluster, and transparently adapt as partitions of
- * data it fetches migrate within the cluster. This client also interacts with the server to allow groups of
- * consumers to load balance consumption using consumer groups (as described below).
+ * This client transparently handles the failure of Kafka brokers, and transparently adapt as topic partitions
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: `adapts`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 2]
```diff
@@ -254,9 +241,28 @@
...
+ * <p>
+ * To avoid this, we will manually commit the offsets only after the corresponding records have been inserted into the
+ * database. This gives us exact control of when a record is considered consumed. This raises the opposite possibility:
+ * the process could fail in the interval after the insert into the database but before the commit (even though this
+ * would likely just be a few milliseconds, it is a possibility). In this case the process that took over consumption
+ * would consume from last committed offset and would repeat the insert of the last batch of data. Used in this way
+ * Kafka provides what is often called "at-least-once" delivery guarantees, as each record will likely be delivered one
+ * time but in failure cases could be duplicated.
+ * <p>
+ * <b>Note: Using automatic offset commits can also give you "at-least-once" delivery, but the requirement is that
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:257]: Sorry for all the additional noise (I decided to tackle a little minor cleanup), but here is the key passage.

**Comment by becketqin** [comment]
LGTM.

**Comment by vahidhashemian** [comment]
LGTM.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 3]
```diff
@@ -141,37 +139,36 @@
...
- *     interval.</li>
+ *     interval. By tuning this value, you may be able to reduce the poll interval, which will reduce the
+ *     impact of group rebalancing.</li>
  * </ol>
  * <p>
- * For use cases where message processing time varies unpredictably, neither of these options may be viable.
+ * For use cases where message processing time varies unpredictably, neither of these options may be sufficient.
  * The recommended way to handle these cases is to move message processing to another thread, which allows
- * the consumer to continue sending heartbeats while the processor is still working. Some care must be taken
+ * the consumer to continue calling {@link #poll(long) poll} the processor is still working. Some care must be taken
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: `while` seems to be missing

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, LGTM, just a couple of minor typos.

**Pull request closed without merging**